doc/scwm.1
.TH scwm 1 "18 March 1999"
.SH NAME
scwm, scwmexec, scwmrepl \- Scheme Configurable Window Manager.
.SH SYNOPSYS
scwm [--display|-d dpy] [--debug][--expression|-e expression] [--file|-f rc_file] [--single-screen|-s] [--blackout|-b] [--version|-V] [--help|-h]
.sp
scwmexec expression
.sp
scwmrepl

.SH DESCRIPTION
.B NOTE: THIS MAN PAGE IS INCOMPLETE!
You should refer to the online reference manual at:

http://www.cs.washington.edu/homes/gjb/scwm-doc/

or use the included `scwm.el' emacs mode and its interactive
documentation features (very nice!).

.B Scwm
is the Scheme Configurable Window Manager.  This is a highly dynamic
and extensible window manager for the X Window System (based
originally on 
.B fvwm2
, but now much enhanced) with Guile Scheme as the
configuration/extension language.  Nearly all decorations can be
changed at run-time or per-window, and eventually many decoration
styles and additional features will be supported through dynamically
loaded code.  A powerful protocol is provided for interacting with the
window manager while it is running. Much more information is available
in the 
.B scwm
info entry.

.sp
The 
.B scwm
command is used to start the window manager itself.
.B scwmexec
is used to pass a single command expression to the window manager, and
prints the resulting output.
.B scwmrepl
starts an interactive session with the window manager.

.SH OPTIONS
.TP
.B "\-D"
.TP
.B "\-\-debug"
Enable lots of debugging messages.

.TP
.B "\-s"
.TP
.B "\-\-single\-screen"
Run only on the first screen of the display.

.TP
.B "\-d"
.TP
.B "\-\-display \fIdpy\fP"
Run on display \fIdpy\fP.

.TP
.B "\-f"
.TP
.B "\-\-file \fIfile\fP"
Read startup commands from the specified \fIfile\fP instead
of \fI.scwmrc\fP or \fIsystem.scwmrc\fP.

.TP
.B "\-e"
.TP
.B "\-\-expression \fIexpr\fP"
Evaluate Scheme expression \fIexpr\fP instead of reading from =
\fI.scwmrc\fP
or \fIsystem.scwmrc\fP.  Multiple \fB\-e\fP
and \fB\-f\fP options may be specified on a single command line and in
this case will be processed in the order in which they were specified.

.TP
.B "\-i"
.TP
.B "\-\-interactive"
Starts \fBscwm\fP in interactive mode \- the \fI.scwmrc\fP
or \fIsystem.scwmrc\fP file is read as usual (unless of course
there are \fB\-f\fP or \fB\-e\fP options.)

.TP
.B "\-b"
.TP
.B "\-\-blackout"
Blacks the screen out to hide the initial capture.

.TP
.B "\-V"
.TP
.B "\-\-version"
Prints the version and exits.
.SH LICENSE
Some files in scwm are derived from fvwm and therefore are subject to
the following fvwm copyright and license terms. However, many files
are entirely new and are under the GNU General Public License as
published by the Free Software Foundation instead. These terms are
believed to be compatible with the GPL, thus the overall distribution
terms are GPL.

\fIfvwm\fP is copyright 1988 by Evans and Sutherland Computer
Corporation, Salt Lake City, Utah, and 1989 by the Massachusetts
Institute of Technology, Cambridge, Massachusetts, All rights
reserved.  It is also copyright 1993 and 1994 by Robert Nation.

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the names of Evans & Sutherland and
M.I.T. not be used in advertising in publicity pertaining to
distribution of the software without specific, written prior
permission.

ROBERT NATION, CHARLES HINES, EVANS & SUTHERLAND, AND M.I.T. DISCLAIM
ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL EVANS &
SUTHERLAND OR M.I.T. BE LIABLE FOR ANY SPECIAL, INDIRECT OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTUOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.

.SH BUGS
See the BUGS file in the scwm distribution for some known ones.
.PP
.SH "SEE ALSO"
.BR guile ( 1 ),
.BR fvwm2 ( 1 ),

.B info
entries for
.RB "`\|" scwm "\|'",
.RB "`\|" guile-ref "\|'",
and
.RB "`\|" Scheme-R4RS "\|'".


.RB "`\|" scwm.el "\|'"s 
scwm-mode for Emacs
.br

Scwm web site:

.B http://scwm.mit.edu/scwm/

Reference manual:

.B http://www.cs.washington.edu/homes/gjb/scwm-doc/
.PP
.SH AUTHORS
Maciej Stachowiak <mstachow@alum.mit.edu>

Greg J. Badros <gjb@cs.washington.edu>
.br
http://www.cs.washington.edu/homes/

and more -- see THANKS file with distribution.


NEWS
Scwm NEWS --- history of user-visible changes.  -*- text -*-
Copyright (C) 1997, 1998, 1999 Greg J. Badros and Maciej Stachowiak
See the end for copying conditions.

Changes for scwm-0.99.6 (not yet released)
* Bug fixes
** menus popped-up from decorations stay on screen
** visibility change hooks are called properly, and during moves/resizes

* window-configuration.scm
** Now changes in global state animate all windows simultaneously, instead of sequentially

* Changed procedures
** Rename menu-global-theme, menu-window-theme to
`make-menu-global-theme' and `make-menu-window-theme' since they are
procs that return menus, not bound directly to menus.

* New primitives
** `window-visibility' for reporting on current visibility state of a window
** `animate-windows' for permitting animations of multiple windows simultaneously
** `cl-resolve-xforms', `cl-reset-dirty-windows' for use in scwm-resolve-hook to permit animation

Changes for scwm-0.99.5.1 (released 26-October-1999)
* Bug fixes
* RPM built to link against libguile.so.6 (guile-1.3.4) instead of
  libguile.so.7 (CVS guile)

Changes for scwm-0.99.5 (released 24-October-1999)

* Bug fixes

* Added interactive procedure specifications

Used to make clear which procedures should be called interactively
via menus or bindings.  Right now just used by prompt-binding, but
will be required of menuitem actions and binding actions in the next
version.

* New configure option, --enable-dump
Rudimentary, not-state-preserving-yet support for guile-hobbit's
unexec capabilities (goal: reduce startup time substantially).

* New preferences options
** *highlight-foreground*, *highlight-background*

* SIGINT (Ctrl-C) is now allowed to interrupt and terminate Scwm
Unless SCWM_CATCH_SIGINT is defined during compilation

* New modules (largely from breaking up flux.scm, factoring system.scwmrc)
** winops-menu.scm
** xlock-menus.scm
** titlebar-buttons.scm
** task-switcher.scm
** synth-events.scm
** shove-window.scm
** session.scm
** rectangle.scm
** lambda-startup-logo.scm
** describe.scm
** basic-styles.scm

* Renamed numerous primitives
** Now (un)shade-window, shaded-window? 
-> was window-(un)shade, window-shaded?
** Standardize on toggle-maximize-{vertical,horizontal,both} instead of random other names
** (de)iconify-window, iconified-window?
-> was (de)iconify, iconified?
** focus-window
-> was focus
** sticky-window?, (un)stick-window
-> was sticky?, (un)stick
** `window-focus-style'
-> was get-window-focus
** `send-button', `send-key'
-> were send-button-press, send-key-press

Changes for scwm-0.99.4 (released 6-October-1999)

* scwm/ subdirectory has been renamed to src/.

* Bug fixes, including gnome icon placement, animated resize/moves, more!

* Minor overhaul of system.scwmrc, rearranged menus, eliminated half-working features, etc.

* system.scwmrc by default now uses the guile-gtk error message window in addition to
  writing errors to stderr (uses changed error-hook)

* Motion events were only dispatching after the double-click timeout;  now they dispatch
  as soon as the motion is greater than a threshold.  Result is more responsive feel on, e.g.,
  titlebar-based window dragging.

* New modules
** last-mapped-window.scm for remembering the last window that got mapped

* Changed modules
** window-configuration.scm now has push-window-configuration, pop-window-configuration
   and window-configuration-menu

* Changed procedures
** X-cut-buffer-string now uses X-fetch-bytes primitive
** X-set-cut-buffer-string! now uses X-store-bytes primitive
** Rename copy-window-title-to-cut-buffer0 to copy-window-title-to-cut-buffer
** Rename paste-window-title-from-cut-buffer0 to paste-window-title-from-cut-buffer

* New primitives
** `X-fetch-bytes', `X-store-bytes' for better primary cut buffer handling. Thanks E. Jay Berkenbilt!

* Changed primitives
** `animated-deiconify-to-last-viewport-position' to `animated-deiconify-to-viewport'
** Drop `animated-deiconify-to-current-viewport'
** Renamed `current-window-with-focus' to `window-with-focus'
** Renamed `current-window-with-pointer' to `window-with-pointer'

* New hooks
** X-ConfigureRequest-hook

* Changed hooks
** X-DestroyNotify-hook now calls with window ID integer, instead of a window object
** error-hook now calls with a single argument, the string error message



Changes for scwm-0.99.3 (released 25-September-1999)

* ViaVoice (IBM TM) module for permitting speech recognition actions

* Turn on backtraces by default (still permit -n option to disable them)
  (guile >= 1.3.2 turned this off)

* Turn on source code position recording by default for line numbers in error messages
  (guile >= 1.3.2 turned this off)

* Constraint features
** Make animated moves and resizes resolve the system throughout the move/resize
   so the constraints are always maintained
** Improved visual constraint representations
** Make constraints submenu to root menu
** Bug fixes in the gtk-toggle-menu

* Improved preferences dialog support, cleaner internally, too.

* New modules
** prompt-binding -- for interrogating and changing key/mouse bindings
** image-path-suggestor -- for using `locate' to find images not in the load-path
** wavplay -- for abstracting different kinds of audio support
** gtk-menu -- for using GTK+'s menus
** prompt-enum -- for choosing among multiple discrete possibilities 
** prompting-shell -- factored out top-level window for various prompting code
** animated-edge-moves -- functions for pushing windows to edges of viewport
** viavoice (C) -- speech recognizer

* New primitives
** `window-decoration-size'
** `window-decoration-ids', `set-window-id-background!'
** `set-window-highlighted-nonant!', `window-highlighted-nonant'
** `set-nonant-highlight-color!', `nonant-highlight-color'
** `focussable-window?'

* Changed primitives
** `lookup-procedure-bindings' takes a new second argument, CONTEXT, to filter returned bindings

utilities/emacs/scwm.el
;;; scwm --- functions for editing and running SCWM code under Emacs

;; Copyright (c) 1998 by Sam Steingold <sds@usa.net>

;; File: <scwm.el - 1998-07-14 Tue 12:44:41 EDT sds@mute.eaglets.com>
;; Author: Sam Steingold <sds@usa.net>
;; Version: $Revision$
;; Keywords: language lisp scheme scwm

;; LCD Archive Entry:
;; scwm|Sam Steingold|sds@usa.net|
;; Functions for editing and running SCWM code under Emacs|
;; $Date$|$Revision$||

;;; $Id$

;;; History:

;; Completion-support added by Greg J. Badros <gjb@cs.washington.edu>
;;    03/11/98 gjb
;;
;; Completition, help and apropos are completely re-worked by sds
;;	1998-03-13 Fri 13:58:16 EST	sds
;;
;; Fixed scwm-run to restart in the same buffer after a crash.
;;	1998-03-17 Tue 15:50:55 EST	sds
;;
;; Made into a major mode.
;;	1998-04-16 Thu 10:38:19 CEST	robbe@orcus.priv.at
;;
;; Added font-lock support for the major mode stuff.
;;	1998-04-16 Thu 17:04:43 EDT	sds
;;
;; Added font-lock support for XEmacs and Emacs19 compatibility.
;;	1998-06-19 Fri 09:28:19 EDT	sds
;;
;; This file is distributed under the GPL. See
;;	<URL:http://www.gnu.ai.mit.edu/copyleft/gpl.html>
;; for further details.

;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with This file; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.

;;; Commentary:

;; Put this file somewhare in your load-path and add the following
;; lines to your .emacs file:

;;   (autoload 'scwm-mode "scwm" "Major mode for editing scwm code. [...]" t)
;;   (setq auto-mode-alist (cons '("scwmrc\\'" . scwm-mode) auto-mode-alist))

;; Now you do M-x scwm-run (C-c C-s) to get the *scwm* buffer, where you
;; can type commands to be sent to scwm; while you can type C-x C-e in
;; your scheme buffers to send the last SEXP to the interpreter.
;; See C-h v scheme-buffer for details on running several scheme
;; interpreters at once.

;; *Alternatively* (that's OR, not XOR) you can type C-j evaluate the
;; last SEXP and insert the results at point, or type C-x C-j to display
;; them in the minibuffer.  This functionality does not require
;; preliminary M-x scwm-run.  Note that you can find your recent
;; minibuffer messages in the buffer *Messages*.  Help and Apropos are
;; also available: type C-h C-a for apropos and C-h C-s for documentation.
;; Type M-TAB to complete symbol at point.

(eval-and-compile
 (or (and (fboundp 'cadr) (fboundp 'unless)) (require 'cl))
 (or (fboundp 'apropos-mode) (autoload 'apropos-mode "apropos"))
 (or (fboundp 'ignore-errors) (autoload 'ignore-errors "cl-macs" nil nil t)))
(eval-when-compile
 (unless (fboundp 'with-output-to-string)
   (defmacro with-output-to-string (&rest body)
     "Execute BODY, return the text it sent to `standard-output', as a string."
     `(let ((standard-output (get-buffer-create (generate-new-buffer-name
                                                 " *string-output*"))))
       ,@body
       (save-excursion
         (set-buffer standard-output)
         (prog1 (buffer-string)
           (kill-buffer standard-output)))))))

;;; Code:

;; user variables
;; ---- ---------

(defvar scwm-repl "scwmrepl" "The path to scwmrepl.")
(defvar scwm-exec "scwmexec" "The path to scwmexec.")

;; Use scheme major mode
(require 'scheme)

;;; XEmacs doesn't have thingatpt. Too bad.
(unless (fboundp 'thing-at-point)
  ;; pacify the compiler (XEmacs only)
  (eval-and-compile (autoload 'id-select-symbol "id-select"))
  (defun thing-at-point (what)
    "Return the thing at point (crippled: symbols only!)."
    (unless (eq what 'symbol)
      (error "crippled `thing-at-point' - symbols only"))
    (let ((zz (id-select-symbol (point))))
      (buffer-substring-no-properties (car zz) (cdr zz)))))

;; user functions
;; ---- ---------

;;;###autoload
(define-derived-mode scwm-mode scheme-mode "Scwm"
  "Major mode for editing scwm code.
Special commands:
\\{scwm-mode-map}
Turning on Scwm mode calls the value of the variable `scwm-mode-hook',
if that value is non-nil.
If you are using Emacs 20.2 or earlier and want to use fontifications,
you have to (require 'font-lock) first.  Sorry.")

(define-key scwm-mode-map [(control j)] 'scwm-eval-print)
(define-key scwm-mode-map [(control c) (control s)] 'scwm-run)
(define-key scwm-mode-map [(control x) (control j)] 'scwm-eval-to-minibuffer)
(define-key scwm-mode-map [(control h) (control s)] 'scwm-documentation)
(define-key scwm-mode-map [(control h) (control a)] 'scwm-apropos)
(define-key scwm-mode-map [(control h) (control f)]
  'scwm-goto-guile-procedure-node)
(define-key scwm-mode-map [(meta tab)] 'scwm-complete-symbol-insert)

;;;###autoload
(defun scwm-run ()
  "Run scwm interaction or pop to an existing one.
Use \\[scheme-send-last-sexp] to eval the last sexp there."
  (interactive)
  (unless (fboundp 'inferior-scheme-mode)
    (let ((ff (symbol-function 'run-scheme)))
      (if (and (consp ff) (eq (car ff) 'autoload)) (load (cadr ff))
	  (error "no `inferior-scheme-mode' and no place to get it from."))))
  (pop-to-buffer (setq scheme-buffer (make-comint "scwm" scwm-repl)))
  (inferior-scheme-mode))

(defsubst scwm-eval (sexp out)
  "Evaluate the SEXP with scwm-exec and print the results to OUT."
  (call-process scwm-exec nil out nil sexp))

(defsubst scwm-eval-last (out)
  "Evaluate the last sexp with scwm-exec and print the results to OUT."
  (scwm-eval (buffer-substring-no-properties
	      (point) (save-excursion (backward-sexp) (point))) out))

(defsubst scwm-safe-call (func args out)
  "Call FUNC with ARGS and output to OUT, checking existence of FUNC first."
  (scwm-eval (concat "(if (bound? " func ") (" func " " args ") "
                     "(display \"This Guile version lacks `" func "'.\n\"))")
             out))

;;;###autoload
(defun scwm-eval-print ()
  "Evaluate the last SEXP and insert the result into the current buffer."
  (interactive) (newline-and-indent)
  (scwm-eval-last t) (newline))

;;;###autoload
(defun scwm-eval-to-minibuffer ()
  "Evaluate the last SEXP and show the result in the minibuffer."
  (interactive)
  (let ((last (buffer-substring-no-properties
	       (point) (save-excursion (backward-sexp) (point)))))
    (message (with-output-to-string (scwm-eval last standard-output)))))

(defalias 'advertised-xscheme-send-previous-expression
    'scwm-eval-to-minibuffer)

;; completion
;; ----------

(defvar scwm-obarray nil "The obarray for scwm completion.")
(defvar scwm-history nil "The input history of SCWM completions.")

(defun scwm-make-obarray ()
  "Create and return an obarray of SCWM symbols."
  ;; can't use read-from-string because "? " is read as 32
  (let ((obarray (make-vector 67 0)) (pos 2)
	(tb (get-buffer-create " *scwm-obarray*")))
    (set-buffer tb) (erase-buffer)
    (scwm-safe-call "apropos-internal" "\".*\"" tb)
    (goto-char 1)
    (while (search-forward " " nil t)
      (intern (buffer-substring-no-properties pos (1- (point))) obarray)
      (setq pos (point)))
    obarray))

(defun scwm-complete-symbol (&optional sym)
  "Complete the current symbol or SYM by querying scwm using apropos-internal.
Returns a string."
  (setq sym (or sym (thing-at-point 'symbol))
	scwm-obarray (or scwm-obarray (scwm-make-obarray)))
  ;; removed the lst arg, `sym', for backward compatibility with e19.
  (completing-read "SCWM symbol: " scwm-obarray nil nil sym scwm-history))

;;;###autoload
(defun scwm-complete-symbol-insert ()
  (interactive)
  (let* ((end (point)) (beg (save-excursion (backward-sexp) (point)))
	 (pat (buffer-substring-no-properties beg end))
	 (comp (try-completion pat scwm-obarray)))
    (cond ((eq comp t) (message "`%s' is complete" pat))
	  ((null comp) (message "Cannot complete `%s'" pat) (ding))
	  ((not (string= comp pat)) (delete-region beg end) (insert comp))
	  (t (message "Making completion list...")
	     (with-output-to-temp-buffer "*Completions*"
	       (display-completion-list (all-completions pat scwm-obarray)))
	     (message "Making completion list...done")))))


;; fontifications
;; --------------
(cond ((boundp 'running-xemacs)
       (put 'scwm-mode 'font-lock-defaults
            (get 'scheme-mode 'font-lock-defaults)))
      ((and (string-lessp emacs-version "20.3")
            (boundp 'font-lock-defaults-alist))
       (unless (assq 'scwm-mode font-lock-defaults-alist)
         (setq font-lock-defaults-alist
               (cons (cons 'scwm-mode
                           (cdr (assq 'scheme-mode font-lock-defaults-alist)))
                     font-lock-defaults-alist)))))

;; help
;; ----

;;;###autoload
(defun scwm-documentation (pat)
  "Query scwm for documentation for the symbol PAT."
  (interactive (list (scwm-complete-symbol)))
  (with-output-to-temp-buffer "*Help*"
    (princ "SCWM documentation `") (princ pat) (princ "':\n\n")
    (scwm-safe-call "procedure-documentation" pat standard-output)
    (set-buffer standard-output) (help-mode)))

;;;###autoload
(defun scwm-apropos (pat)
  "List all scwm symbols matching PAT."
  ;; (interactive (interactive-token "SCWM Apropos"))
  (interactive
   (list (read-string
	  "SCWM Apropos: "
	  (format "%s" (or (thing-at-point 'symbol)
			   (progn (skip-syntax-backward "^w")
				  (if (bobp) () (backward-char 1))
				  (thing-at-point 'symbol)) "")))))
  (with-output-to-temp-buffer "*Apropos*"
    (princ "SCWM apropos `") (princ pat) (princ "':\n\n")
    (scwm-safe-call "apropos" (concat "\"" pat "\"") standard-output)
    (set-buffer standard-output) (apropos-mode)))

;; info interface
(defvar scwm-info-file-list
  '(("r4rs" . "Index")
    ("guile-ref" . "Procedure Index")
    ("guile-ref" . "Variable Index")
    ("scwm" . "Function Index")
    ("scwm" . "Variable Index"))
  "AssocList of Info files that describe Guile procedures.
An element is of the form (FILE . NODE).")

(defun scwm-find-guile-procedure-nodes (procedure)
  "Return a list of locations documenting PROCEDURE.
The variable `scwm-info-file-list' defines heuristics for which Info
manual to try.  The locations are of the format used in Info-history,
i.e. (FILENAME NODENAME BUFFERPOS)"
  (let ((where '())
	(cmd-desc (concat "^\\* " (regexp-quote procedure)
			  ":\\s *\\(.*\\)\\.$"))
	(file-list scwm-info-file-list))
    (save-window-excursion
      (save-excursion
        (while file-list
          (let ((file (car (car file-list)))
                (index (cdr (car file-list))))
            (setq file-list (cdr file-list))
            (ignore-errors
              (Info-find-node file index)
              ;; Take the index node off the Info history.
              (setq Info-history (cdr Info-history))
              (goto-char (point-max))
              (while (re-search-backward cmd-desc nil t)
                (setq where (cons (list Info-current-file
                                        (buffer-substring
                                         (match-beginning 1)
                                         (match-end 1))
                                        0)
                                  where))))))))
    where))

;;;###autoload
(defun scwm-goto-guile-procedure-node (procedure)
  "Go to the Info node in the Guile manual for procedure PROCEDURE.
The procedure is found by looking up in the Guile Reference manual's
Procedure Index or in another manual found via the variable
`scwm-info-file-list'."
  (interactive (list (scwm-complete-symbol)))
  (let ((where (scwm-find-guile-procedure-nodes procedure)))
    (unless where (error "Couldn't find documentation for %s" procedure))
    (let ((num-matches (length where)))
      ;; Get Info running, and pop to it in another window.
      (save-window-excursion (info))
      (pop-to-buffer "*info*")
      ;; (filename nodename bufferpos)
      (Info-find-node (car (car where)) (car (cdr (car where))))
      (when (> num-matches 1)
        ;; Info-find-node already pushed (car where) onto
        ;; Info-history.  Put the other nodes that were found on
        ;; the history.
        (setq Info-history (nconc (cdr where) Info-history))
        (message "Found %d other entr%s.  Use %s to see %s."
                 (1- num-matches) (if (> num-matches 2) "ies" "y")
                 (substitute-command-keys "\\[Info-last]")
                 (if (> num-matches 2) "them" "it"))))))

(provide 'scwm)
;;; scwm ends here
